File: Application/src/com/hhp227/application/ReplyModifyActivity.java
package com.hhp227.application;

import java.util.HashMap;
import java.util.Map;

import org.json.JSONException;
import org.json.JSONObject;

import android.app.ActionBar;
import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.Toast;

import com.android.volley.AuthFailureError;
import com.android.volley.Request;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.StringRequest;
import com.hhp227.application.app.AppController;
import com.hhp227.application.app.URLs;
import com.hhp227.application.user.User;

public class ReplyModifyActivity extends Activity {
    private static final String TAG = "수정화면";

    private static int reply_id;
    private static String replymsg;
    private EditText inputreply;
    private ProgressDialog progressDialog;
    private ListView listView;
    private View headerView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_reply_modify);

        ActionBar actionBar = getActionBar();
        actionBar.setDisplayHomeAsUpEnabled(true);
        actionBar.setDisplayShowHomeEnabled(false);
        listView = (ListView) findViewById(R.id.lv_write);
        headerView = getLayoutInflater().inflate(R.layout.input_text, null, false);
        inputreply = (EditText) headerView.findViewById(R.id.inputcontents);
        listView.addHeaderView(headerView);

        listView.setAdapter(null);

        progressDialog = new ProgressDialog(this);
        progressDialog.setCancelable(false);

        Intent intent = getIntent();
        reply_id = intent.getIntExtra("reply_id", reply_id);
        replymsg = intent.getStringExtra("replymsg");

        inputreply.setText(replymsg);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.write, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();
        switch(id) {
            case android.R.id.home :
                finish();
                return true;
            case R.id.action_send :
                String text = inputreply.getText().toString().trim();

                if(!text.isEmpty()) {
                    Action_Send(text);
                } else {
                    Toast.makeText(getApplicationContext(), "내용을 입력하세요.", Toast.LENGTH_LONG).show();
                }
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    /**
     * 전송을 눌렀을때 작동하는 내용
     */
    private void Action_Send(final String text) {
        String tag_string_req = "req_send";
        progressDialog.setMessage("전송중...");
        showProgressDialog();

        String URL_REPLY = URLs.URL_REPLY.replace("{REPLY_ID}", String.valueOf(reply_id));
        StringRequest stringRequest = new StringRequest(Request.Method.PUT, URL_REPLY, new Response.Listener<String>() {
            @Override
            public void onResponse(String response) {
                Log.e(TAG, "전송 응답 : " + response);
                hideProgressDialog();
                try {
                    JSONObject jsonObject = new JSONObject(response);
                    boolean error = jsonObject.getBoolean("error");
                    if(!error) {
                        String message = jsonObject.getString("message");
                        
                        modifyData(text); // 클라이언트측에 처리해줘야할 메소드
                        Toast.makeText(getApplicationContext(), "수정완료", Toast.LENGTH_LONG).show();
                    } else {
                        String errorMsg = jsonObject.getString("message");
                        Toast.makeText(getApplicationContext(), "수정할 수 없습니다.", Toast.LENGTH_LONG).show();
                    }
                } catch(JSONException e) {
                    Log.e(TAG, "JSON에러" + e);
                }
            }
        }, new Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError error) {
                Log.e(TAG, "전송 에러 : " + error.getMessage());
                Toast.makeText(getApplicationContext(), error.getMessage(), Toast.LENGTH_LONG).show();
                hideProgressDialog();
            }
        }) {
            User user = AppController.getInstance().getPrefManager().getUser();

            String apikey = user.getApi_key();
            @Override
            public Map<String, String> getHeaders() throws AuthFailureError {
                HashMap<String, String> headers = new HashMap<String, String>();
                headers.put("Authorization", apikey);
                return headers;
            }
            @Override
            protected Map<String, String> getParams() {
                Map<String, String> params = new HashMap<String, String>();
                params.put("reply", text);
                params.put("status", "0");

                return params;
            }
        };
        AppController.getInstance().addToRequestQueue(stringRequest, tag_string_req);
    }

    /**
     * 전송이 완료되면 이전 Activity에 값을 반환
     */
    private void modifyData(String reply) {
        // 입력 자판 숨기기
        View view = this.getCurrentFocus();
        if(view != null) {
            InputMethodManager inputMethodManager = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
            inputMethodManager.hideSoftInputFromWindow(view.getWindowToken(), 0);
        }
        Intent intent = new Intent(this, DetailActivity.class);
        intent.putExtra("modify_reply", reply);
        setResult(RESULT_OK, intent);
        finish();
    }

    private void showProgressDialog() {
        if(!progressDialog.isShowing())
            progressDialog.show();
    }

    private void hideProgressDialog() {
        if(progressDialog.isShowing())
            progressDialog.dismiss();
    }
}

File: app/src/main/java/com/hhp227/application/activity/ReplyModifyActivity.kt
package com.hhp227.application.activity

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.*
import android.view.inputmethod.InputMethodManager
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.flowWithLifecycle
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.RecyclerView
import com.hhp227.application.R
import com.hhp227.application.databinding.ActivityReplyModifyBinding
import com.hhp227.application.databinding.InputTextBinding
import com.hhp227.application.dto.ReplyItem
import com.hhp227.application.viewmodel.ReplyModifyViewModel
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach

class ReplyModifyActivity : AppCompatActivity() {
    private val viewModel: ReplyModifyViewModel by viewModels()

    private lateinit var binding: ActivityReplyModifyBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityReplyModifyBinding.inflate(layoutInflater)
        binding.recyclerView.adapter = object : RecyclerView.Adapter<ItemHolder>() {
            override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemHolder = ItemHolder(InputTextBinding.inflate(layoutInflater))

            override fun getItemCount(): Int = 1

            override fun onBindViewHolder(holder: ItemHolder, position: Int) {
                holder.bind(viewModel.replyItem)
            }
        }
        viewModel.replyItem = intent.getParcelableExtra("reply") ?: ReplyItem.Reply()
        viewModel.position = intent.getIntExtra("position", 0)

        setContentView(binding.root)
        setSupportActionBar(binding.toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        viewModel.state.flowWithLifecycle(lifecycle, Lifecycle.State.STARTED).onEach { state ->
            when {
                state.isLoading -> {

                }
                state.text != null -> {
                    val intent = Intent(this, PostDetailActivity::class.java)
                        .putExtra("reply", state.text)
                        .putExtra("position", viewModel.position)

                    setResult(RESULT_OK, intent)
                    finish()
                    currentFocus?.let {
                        (getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager).hideSoftInputFromWindow(it.windowToken, 0)
                    }
                }
                state.error.isNotBlank() -> {
                    Toast.makeText(this, state.error, Toast.LENGTH_LONG).show()
                }
            }
        }.launchIn(lifecycleScope)
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.write, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean = when (item.itemId) {
        android.R.id.home -> {
            onBackPressed()
            true
        }
        R.id.actionSend -> {
            val text = (binding.recyclerView.getChildViewHolder(binding.recyclerView.getChildAt(0)) as ItemHolder).binding.etText.text.toString()

            viewModel.updateReply(text)
            true
        }
        else -> super.onOptionsItemSelected(item)
    }

    inner class ItemHolder(val binding: InputTextBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(replyItem: ReplyItem.Reply) {
            binding.etText.setText(replyItem.reply)
        }
    }
}